### PR TITLE
Add fragment presenters

### PR DESCRIPTION
**Pull request by Valentindp**

### CODE REVIEW

File: app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    override fun init() {
+        loadPhoto()
+        updateTextSelectedButton()
+        updateColorSelectedButton()
+        updateSiteTextView()
+        updateNameTextView()
+        updateDescriptionTextView()
+        updateCountryTextView()
+        view.setOnClickListenerSelectedButton()
```
Comment (severianremi) [inline app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt]: view.setOnClickListenerSelectedButton() - this logic should be in View(Activity/Fragment)

File: app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    override fun onFinish() {
+        val intent = Intent(view as CityActivity, MainActivity::class.java)
+        intent.putExtra(CityActivity.CITY_KEY, element)
+        (view as CityActivity).setResult(AppCompatActivity.RESULT_OK, intent)
```
Comment (severianremi) [inline app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt]: It's not really nice when you cast to the CityActivity and handle such logic into the presenter.
The presenter should handle a request, calculation, but not clicks, startActivities()...

File: app/src/main/java/valentyn/androidtasks/presenters/CityFragmentPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    private val dataset: List<City> = CityRepository.dataCitys
+
+    private fun onCityClicked(city: City) {
+        val intent = Intent(fragment.activity, CityActivity::class.java)
+        intent.putExtra(CityActivity.CITY_KEY, city)
+        fragment.startActivityForResult(intent, 1)
+    }
+
+    fun getRecyclerViewAdapter(): CitysAdapter = CitysAdapter(dataset) { city: City -> onCityClicked(city) }
```
Comment (severianremi) [inline app/src/main/java/valentyn/androidtasks/presenters/CityFragmentPresenter.kt]: Your adapter should be created in the Activity/Fragment

File: app/src/main/java/valentyn/androidtasks/presenters/CityFragmentPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import valentyn.androidtasks.views.CityActivity
+
+class CityFragmentPresenter(private val fragment: CityFragment) {
+
+    private val dataset: List<City> = CityRepository.dataCitys
+
+    private fun onCityClicked(city: City) {
+        val intent = Intent(fragment.activity, CityActivity::class.java)
+        intent.putExtra(CityActivity.CITY_KEY, city)
+        fragment.startActivityForResult(intent, 1)
```
Comment (severianremi) [inline app/src/main/java/valentyn/androidtasks/presenters/CityFragmentPresenter.kt]: It should be in view

File: app/src/main/java/valentyn/androidtasks/presenters/ElementPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        view?.updateTextSelectedButton(element?.getTextSelected()!!)
+    }
+
+    override fun updateColorSelectedButton() {
+        view?.updateColorSelectedButton(element?.getColorSelected()!!)
+    }
+
+    override fun onAttach(view: ElementContract.View, id: Long?, key: String) {
+        this.view = view
+        this.element = RealmRepository.getRealmObject(id, key)
```
Comment (severianremi) [inline app/src/main/java/valentyn/androidtasks/presenters/ElementPresenter.kt]: It's really not a good idea to save an element as a variable. The best way to get an element from the database and then update UI.

File: app/src/main/java/valentyn/androidtasks/repository/RealmManager.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package valentyn.androidtasks.repository
+
+import io.realm.Realm
+
+object RealmManager {
+
+    private val realmList = ThreadLocal<Realm>()
+
+    fun getRealm(): Realm? {
+        var realm = realmList.get()
```
Comment (severianremi) [inline app/src/main/java/valentyn/androidtasks/repository/RealmManager.kt]: You do not need this here!!

Comment (severianremi) [reply app/src/main/java/valentyn/androidtasks/repository/RealmManager.kt]: The best way to work with the realm is each time Realm.getDefaultInstance() and then open transaction after that close transaction.

File: app/src/main/java/valentyn/androidtasks/repository/RealmRepository.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+object RealmRepository : Storage<RealmObject> {
+
+    override fun save(element: RealmObject) {
+        val realm = RealmManager.getRealm()
+        realm?.executeTransaction { realm.copyToRealmOrUpdate(element) }
+    }
+
+    fun getRealmObject(id: Long?, key: String): BaseContract.Model? = when (key) {
+        CityActivity.CITY_KEY -> RealmManager.getRealm()?.where<City>()?.equalTo("id", id)?.findFirst()
```
Comment (severianremi) [inline app/src/main/java/valentyn/androidtasks/repository/RealmRepository.kt]: You can get an exception when you work in the wrong thread with Realm. The best practice not to work with Realm on the main thread

File: app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt (hunk 1 above)
Comment (Valentindp) [reply app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt]: This is the old code. In the new version setOnClickListenerSelectedButton is called in the view.
Like this:
_selectedButton.setOnClickListener { presenter.setOnClickListenerSelectedButton() }_

File: app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt (hunk 2 above)
Comment (Valentindp) [reply app/src/main/java/valentyn/androidtasks/presenters/CityActivityPresenter.kt]: This is the old version (

**Pull request merged**